[MRG+1] BUG: Fix fetch_kddcup99 for Python 3

nmayorov opened the pull request:
This is a small fix to account for str / bytes issues in `fetch_kddcup99`.

Commit:
BUG: Fixed fetch_kddcup99 for Python 3

BUG: Fixed comparison with bytes in kddcup99.py + test

MAINT: Changed default 'percent10' to True in fetch_kddcup99

rvraghav93 commented:
LGTM Thanks for the  fix :)

nmayorov commented:
OK, I changed to MRG+1

amueller commented:
why was this not caught by a test? because the tests don't download? But if the file is downloaded already, the tests should run, right?

rvraghav93 commented:
Ah there are no tests for `kddcup99.py`. @nmayorov could you introduce a test for this?

nmayorov commented:
@rvraghav93 I will add a test similar to `test_covtype.py`.

rvraghav93 commented:
yes thanks :)

nmayorov commented:
@rvraghav93 not sure how to test it, `test_covtype.py` doesn't download data during testing (the same for other `fetch` test routines). 

The problem with `fetch_kddcup99` occurs during the download and parsing phase. So what should I test? The download part doesn't seem to fit into unit testing.

Could someone just merge it and be done with it?

rvraghav93 commented:
I guess yes... @amueller merge?

amueller commented:
Wait, there are still tests for the other downloaded datasets, right? If the data was downloaded, then the tests are run with the downloaded data, otherwise they are mocked, right?
If not, we should make this more consistent.

nmayorov commented:
I added a test (even though it doesn't target this fix), and it was helpful to find a related bug with comparison (so @amueller was right). 

I noticed other strange thing: the current implementation with `subset='SF', 'http' or 'smtp'`returns data with 4, 3 and 3 features respectively (and from the code it seems like it was intended). I'm not well familiar with this dataset, but it looks strange and contradicts the docstring (for now I modified the docstring).

I think we need to ask the author @ngoix of `kddcup.py` what does it mean, is it right or wrong?

ngoix commented:
Yes you are right the docstring was wrong.

ngoix commented on sklearn/datasets/tests/test_kddcup99.py:
```diff
@@ -0,0 +1,48 @@
+    data = fetch_kddcup99('http', percent10=True)
+    assert_equal(data.data.shape, (58725, 3))
+    assert_equal(data.target.shape, (58725,))
+    print(data.data.shape, data.target.shape)
```
print to be removed

ngoix commented on sklearn/datasets/tests/test_kddcup99.py:
```diff
@@ -0,0 +1,48 @@
+    data = fetch_kddcup99('smtp', percent10=True)
+    assert_equal(data.data.shape, (9571, 3))
+    assert_equal(data.target.shape, (9571,))
+    print(data.data.shape, data.target.shape)
```
idem

ngoix commented:
+1 for merge

nmayorov commented:
@ngoix thanks, 

I also changed "Load and return the kddcup 99 dataset (regression)" to "Load and return the kddcup 99 dataset (anomaly detection)". Is it appropriate description (calling it "regression" was definitely misleading)?

ngoix commented:
Yes right thanks! You could also call it more generally classification.

nmayorov commented:
Changed to standard "classification".

amueller commented on sklearn/datasets/tests/test_kddcup99.py:
```diff
@@ -0,0 +1,46 @@
+    assert_equal(data.target.shape, (9571,))
+
+
+if __name__ == '__main__':
```
please remove this. the file should be run using nostests.

amueller commented:
+1 apart from the ``__main__``...

nmayorov commented:
Done.

nmayorov commented:
Another thing: how about changing default of `percent10` to `True`? The full data is big, and perhaps requires more than 8GB RAM to be fetched without issues (all data is stored as Python list at some point).

agramfort commented:
@ngoix any objection?

ngoix commented:
No I'm ok with this change!

agramfort commented:
+1 for merge when @amueller is happy

amueller commented:
squash maybe? LGTM.

nmayorov commented:
@amueller, squashed.

agramfort commented:
thanks @nmayorov
